dGame/dCinema/Recorder.cpp

```diff
@@ -0,0 +1,1195 @@
+			// Skip the player proximity record
+			ActingDispatch(actor, records, index + 1, variables);
+			return;
+		}
+
+		auto actionTaken = std::make_shared<bool>(false);
+
+		PlayerProximityDispatch(actor, records, index, variables, actionTaken);
+
+		if (playerProximityRecord->timeout <= 0.0f) {
+			return;
+		}
+
+		Game::entityManager->GetZoneControlEntity()->AddCallbackTimer(playerProximityRecord->timeout, [playerProximityRecord, actor, records, index, variables, actionTaken]() {
+			if (*actionTaken) {
+				return;
+			}
+
+			*actionTaken = true;
+
+			for (auto i = 0; i < records.size(); ++i) {
+				auto* record = records[i];
+
+				if (record->m_Name == playerProximityRecord->timeoutLabel) {
+					ActingDispatch(actor, records, i, variables);
+					return;
+				}
+			}
+		});
+
+		return;
+	}
+
+	// Check if the record is a coroutine record
+	auto* coroutineRecord = dynamic_cast<CoroutineRecord*>(record);
+
+	if (coroutineRecord) {
+		actor->AddCallbackTimer(delay, [actor, coroutineRecord, index, variables]() {
+			ActingDispatch(actor, coroutineRecord->records, 0, variables);
+		});
+	}
+
+	// Check if the record is a path find record
+	auto* pathFindRecord = dynamic_cast<PathFindRecord*>(record);
+
+	if (pathFindRecord) {
+		auto* movementAiComponent = actor->GetComponent<MovementAIComponent>();
+
+		if (movementAiComponent == nullptr) {
+			movementAiComponent = actor->AddComponent<MovementAIComponent>(MovementAIInfo{});
+		}
+		
+		movementAiComponent->SetDestination(pathFindRecord->position);
+		movementAiComponent->SetMaxSpeed(pathFindRecord->speed);
+		
+		PathFindDispatch(actor, records, index, variables);
+
+		return;
+	}
+
+	actor->AddCallbackTimer(delay, [actor, records, index, variables]() {
+		ActingDispatch(actor, records, index + 1, variables);
+	});
+
+	record->Act(actor);
+}
+
+void Cinema::Recording::Recorder::PlayerProximityDispatch(Entity* actor, const std::vector<Record*>& records, size_t index, Play* variables, std::shared_ptr<bool> actionTaken) {
+	auto* record = dynamic_cast<PlayerProximityRecord*>(records[index]);
+	auto* player = Game::entityManager->GetEntity(variables->player);
+
+	if (player == nullptr || record == nullptr) {
+		return;
+	}
+
+	const auto& playerPosition = player->GetPosition();
+	const auto& actorPosition = actor->GetPosition();
+
+	const auto distance = NiPoint3::Distance(playerPosition, actorPosition);
+
+	if (distance <= record->distance) {
+		if (*actionTaken) {
+			return;
+		}
+
+		*actionTaken = true;
+
+		ActingDispatch(actor, records, index + 1, variables);
+
+		return;
+	}
+	
+	Game::entityManager->GetZoneControlEntity()->AddCallbackTimer(1.0f, [actor, records, index, variables, actionTaken]() {
+		PlayerProximityDispatch(actor, records, index, variables, actionTaken);
+	});
+}
+
+void Cinema::Recording::Recorder::PathFindDispatch(Entity* actor, const std::vector<Record*>& records, size_t index, Play* variables) {
+	auto* record = dynamic_cast<PathFindRecord*>(records[index]);
```

Comment:
Same thing here, this should be abstracted into overloaded methods so you can just call those on the object instead of needing to do this cast and then check the ptr